libbpf-tools/opensnoop.c

```diff
@@ -0,0 +1,298 @@
+	return vfprintf(stderr, format, args);
+}
+
+static int bump_memlock_rlimit(void)
+{
+	struct rlimit rlim_new = {
+		.rlim_cur	= RLIM_INFINITY,
+		.rlim_max	= RLIM_INFINITY,
+	};
+
+	return setrlimit(RLIMIT_MEMLOCK, &rlim_new);
+}
+
+void handle_event(void *ctx, int cpu, void *data, __u32 data_sz)
+{
+	const struct event *e = data;
+	struct tm *tm;
+	char ts[32];
+	time_t t;
+	int fd, err;
+
+	/* name filtering is currently done in user space */
+	if (env.name && strstr(e->comm, env.name) == NULL)
+		return;
+
+	/* prepare fields */
+	time(&t);
+	tm = localtime(&t);
+	strftime(ts, sizeof(ts), "%H:%M:%S", tm);
+	if (e->ret >= 0) {
+		fd = e->ret;
+		err = 0;
+	} else {
+		fd = -1;
+		err = - e->ret;
+	}
+
+	/* print output */
+	if (env.timestamp)
+		printf("%-8s ", ts);
+	if (env.print_uid)
+		printf("%-6d ", e->uid);
+	printf("%-6d %-16s %3d %3d ", e->pid, e->comm, fd, err);
+	if (env.extended)
+		printf("%08o ", e->flags);
+	printf("%s\n", e->fname);
+}
+
+void handle_lost_events(void *ctx, int cpu, __u64 lost_cnt)
+{
+	printf("Lost %llu events on CPU #%d!\n", lost_cnt, cpu);
+}
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct perf_buffer_opts pb_opts;
+	struct perf_buffer *pb = NULL;
+	struct opensnoop_bpf *obj;
+	struct timeval time_now = {}, time_duration = {}, time_end = {};
+	int err;
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	libbpf_set_print(libbpf_print_fn);
+
+	err = bump_memlock_rlimit();
+	if (err) {
+		fprintf(stderr, "failed to increase rlimit: %d\n", err);
+		return 1;
+	}
+
+	obj = opensnoop_bpf__open();
+	if (!obj) {
+		fprintf(stderr, "failed to open and/or load BPF object\n");
+		return 1;
+	}
+
+	/* initialize global data (filtering options) */
+	obj->rodata->targ_tgid = env.pid;
+	obj->rodata->targ_pid = env.tid;
+	obj->rodata->targ_uid = env.uid;
+	obj->rodata->targ_failed = env.failed;
+
+	err = opensnoop_bpf__load(obj);
+	if (err) {
+		fprintf(stderr, "failed to load BPF object: %d\n", err);
+		goto cleanup;
+	}
+
+	err = opensnoop_bpf__attach(obj);
+	if (err) {
+		fprintf(stderr, "failed to attach BPF programs\n");
+		goto cleanup;
+	}
+
+	/* print headers */
+	if (env.timestamp)
+		printf("%-8s ", "TIME");
+	if (env.print_uid)
+		printf("%-6s ", "UID");
+	printf("%-6s %-16s %3s %3s ", "PID", "COMM", "FD", "ERR");
+	if (env.extended)
+		printf("%-8s ", "FLAGS");
+	printf("%s\n", "PATH");
+
+	/* setup event callbacks */
+	pb_opts.sample_cb = handle_event;
+	pb_opts.lost_cb = handle_lost_events;
+	pb = perf_buffer__new(bpf_map__fd(obj->maps.events), 64, &pb_opts);
+	err = libbpf_get_error(pb);
+	if (err) {
+		pb = NULL;
+		fprintf(stderr, "failed to open perf buffer: %d\n", err);
+		goto cleanup;
+	}
+
+	/* setup duration */
+	if (env.duration) {
+		gettimeofday(&time_now, NULL);
+		time_duration.tv_sec = env.duration;
+		timeradd(&time_now, &time_duration, &time_end);
```

Comment:
nit: `clock_gettime(CLOCK_MONOTONIC, &value)` might be a better way to measure passage of time?..